### ISSUE TITLE
'OrthoPanel' object has no attribute '_OrthoPanel__xcanvas'

### ISSUE THREAD

**Issue by dmd** [pauldmccarthy/fsleyes#78]
Running under [x2go](https://wiki.x2go.org/doku.php), fsleyes/FSLeyes version 1.0.5 dies on startup with:

```

Traceback (most recent call last):
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/gl/__init__.py", line 690, in create
    ready()
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/main.py", line 574, in realCallback
    callback()
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/main.py", line 370, in buildGui
    [shutdown])
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/main.py", line 785, in makeFrame
    closeHandlers=closeHandlers)
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/frame.py", line 310, in __init__
    self.__restoreState(restore)
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/frame.py", line 1440, in __restoreState
    layouts.loadLayout(self, 'default')
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/layouts.py", line 104, in loadLayout
    applyLayout(frame, name, layout, **kwargs)
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/layouts.py", line 143, in applyLayout
    frame.addViewPanel(vp, defaultLayout=False)
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/frame.py", line 511, in addViewPanel
    panel = panelCls(self.__mainPanel, self.__overlayList, childDC, self)
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/views/orthopanel.py", line 210, in __init__
    zax=0)
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/gl/wxglslicecanvas.py", line 44, in __init__
    wxgl.GLCanvas          .__init__(self, parent, **attrs)
wx._core.wxAssertionError: C++ assertion ""Assert failure"" failed at /home/conda/feedstock_root/build_artifacts/wxpython_1616628589829/work/ext/wxWidgets/src/unix/glx11.cpp(626) in InitVisual(): Failed to get a XVisualInfo for the requested attributes.
 WARNING              idle.py  579: __idleLoop      - Idle task __bgfgColourChanged crashed - AttributeError: 'OrthoPanel' object has no attribute '_OrthoPanel__xcanvas'
Traceback (most recent call last):
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsl/utils/idle.py", line 576, in __idleLoop
    task.task(*task.args, **task.kwargs)
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/views/canvaspanel.py", line 592, in __bgfgColourChanged
    canvases  = self.getGLCanvases()
  File "/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/views/orthopanel.py", line 518, in getGLCanvases
    canvas = [self.__xcanvas, self.__ycanvas, self.__zcanvas]
AttributeError: 'OrthoPanel' object has no attribute '_OrthoPanel__xcanvas'
```

Here is the output from glxinfo:

```
name of display: :52
display: :52  screen: 0
direct rendering: Yes
server glx vendor string: SGI
server glx version string: 1.2
server glx extensions:
    GLX_ARB_multisample, GLX_EXT_import_context, GLX_EXT_visual_info,
    GLX_EXT_visual_rating, GLX_OML_swap_method, GLX_SGIS_multisample,
    GLX_SGIX_fbconfig, GLX_SGIX_hyperpipe, GLX_SGIX_swap_barrier,
    GLX_SGI_make_current_read
client glx vendor string: Mesa Project and SGI
client glx version string: 1.4
client glx extensions:
    GLX_ARB_context_flush_control, GLX_ARB_create_context,
    GLX_ARB_create_context_profile, GLX_ARB_create_context_robustness,
    GLX_ARB_fbconfig_float, GLX_ARB_framebuffer_sRGB,
    GLX_ARB_get_proc_address, GLX_ARB_multisample, GLX_EXT_buffer_age,
    GLX_EXT_create_context_es2_profile, GLX_EXT_create_context_es_profile,
    GLX_EXT_fbconfig_packed_float, GLX_EXT_framebuffer_sRGB,
    GLX_EXT_import_context, GLX_EXT_texture_from_pixmap, GLX_EXT_visual_info,
    GLX_EXT_visual_rating, GLX_INTEL_swap_event, GLX_MESA_copy_sub_buffer,
    GLX_MESA_multithread_makecurrent, GLX_MESA_query_renderer,
    GLX_MESA_swap_control, GLX_OML_swap_method, GLX_OML_sync_control,
    GLX_SGIS_multisample, GLX_SGIX_fbconfig, GLX_SGIX_pbuffer,
    GLX_SGIX_visual_select_group, GLX_SGI_make_current_read,
    GLX_SGI_swap_control, GLX_SGI_video_sync
GLX version: 1.2
GLX extensions:
    GLX_ARB_get_proc_address, GLX_ARB_multisample, GLX_EXT_import_context,
    GLX_EXT_visual_info, GLX_EXT_visual_rating, GLX_MESA_query_renderer,
    GLX_OML_swap_method, GLX_SGIS_multisample, GLX_SGIX_fbconfig,
    GLX_SGI_make_current_read
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: VMware, Inc. (0xffffffff)
    Device: llvmpipe (LLVM 7.0, 256 bits) (0xffffffff)
    Version: 18.3.4
    Accelerated: no
    Video memory: 128704MB
    Unified memory: no
    Preferred profile: compat (0x2)
    Max core profile version: 0.0
    Max compat profile version: 2.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 2.0
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: llvmpipe (LLVM 7.0, 256 bits)
OpenGL version string: 2.1 Mesa 18.3.4
OpenGL shading language version string: 1.20
OpenGL extensions:
    GL_AMD_conservative_depth, GL_AMD_draw_buffers_blend,
    GL_AMD_multi_draw_indirect, GL_AMD_seamless_cubemap_per_texture,
    GL_AMD_shader_stencil_export, GL_AMD_shader_trinary_minmax,
    GL_ANGLE_texture_compression_dxt3, GL_ANGLE_texture_compression_dxt5,
    GL_APPLE_packed_pixels, GL_ARB_ES2_compatibility,
    GL_ARB_ES3_compatibility, GL_ARB_arrays_of_arrays, GL_ARB_base_instance,
    GL_ARB_blend_func_extended, GL_ARB_buffer_storage,
    GL_ARB_clear_buffer_object, GL_ARB_clear_texture, GL_ARB_clip_control,
    GL_ARB_color_buffer_float, GL_ARB_compressed_texture_pixel_storage,
    GL_ARB_conditional_render_inverted, GL_ARB_conservative_depth,
    GL_ARB_copy_buffer, GL_ARB_copy_image, GL_ARB_cull_distance,
    GL_ARB_debug_output, GL_ARB_depth_buffer_float, GL_ARB_depth_clamp,
    GL_ARB_depth_texture, GL_ARB_draw_buffers, GL_ARB_draw_buffers_blend,
    GL_ARB_draw_elements_base_vertex, GL_ARB_draw_indirect,
    GL_ARB_draw_instanced, GL_ARB_enhanced_layouts,
    GL_ARB_explicit_attrib_location, GL_ARB_explicit_uniform_location,
    GL_ARB_fragment_coord_conventions, GL_ARB_fragment_layer_viewport,
    GL_ARB_fragment_program, GL_ARB_fragment_program_shadow,
    GL_ARB_fragment_shader, GL_ARB_framebuffer_object,
    GL_ARB_framebuffer_sRGB, GL_ARB_get_program_binary,
    GL_ARB_get_texture_sub_image, GL_ARB_half_float_pixel,
    GL_ARB_half_float_vertex, GL_ARB_instanced_arrays,
    GL_ARB_internalformat_query, GL_ARB_internalformat_query2,
    GL_ARB_invalidate_subdata, GL_ARB_map_buffer_alignment,
    GL_ARB_map_buffer_range, GL_ARB_multi_bind, GL_ARB_multi_draw_indirect,
    GL_ARB_multisample, GL_ARB_multitexture, GL_ARB_occlusion_query,
    GL_ARB_occlusion_query2, GL_ARB_pipeline_statistics_query,
    GL_ARB_pixel_buffer_object, GL_ARB_point_parameters, GL_ARB_point_sprite,
    GL_ARB_polygon_offset_clamp, GL_ARB_program_interface_query,
    GL_ARB_provoking_vertex, GL_ARB_robustness, GL_ARB_sampler_objects,
    GL_ARB_seamless_cube_map, GL_ARB_seamless_cubemap_per_texture,
    GL_ARB_separate_shader_objects, GL_ARB_shader_bit_encoding,
    GL_ARB_shader_objects, GL_ARB_shader_stencil_export,
    GL_ARB_shader_texture_lod, GL_ARB_shading_language_100,
    GL_ARB_shading_language_420pack, GL_ARB_shading_language_packing,
    GL_ARB_shadow, GL_ARB_stencil_texturing, GL_ARB_sync,
    GL_ARB_texture_border_clamp, GL_ARB_texture_buffer_object,
    GL_ARB_texture_buffer_object_rgb32, GL_ARB_texture_buffer_range,
    GL_ARB_texture_compression, GL_ARB_texture_compression_bptc,
    GL_ARB_texture_compression_rgtc, GL_ARB_texture_cube_map,
    GL_ARB_texture_cube_map_array, GL_ARB_texture_env_add,
    GL_ARB_texture_env_combine, GL_ARB_texture_env_crossbar,
    GL_ARB_texture_env_dot3, GL_ARB_texture_gather,
    GL_ARB_texture_mirror_clamp_to_edge, GL_ARB_texture_mirrored_repeat,
    GL_ARB_texture_multisample, GL_ARB_texture_non_power_of_two,
    GL_ARB_texture_query_levels, GL_ARB_texture_query_lod,
    GL_ARB_texture_rectangle, GL_ARB_texture_rg, GL_ARB_texture_rgb10_a2ui,
    GL_ARB_texture_stencil8, GL_ARB_texture_storage,
    GL_ARB_texture_storage_multisample, GL_ARB_texture_swizzle,
    GL_ARB_texture_view, GL_ARB_timer_query, GL_ARB_transform_feedback2,
    GL_ARB_transform_feedback3, GL_ARB_transform_feedback_instanced,
    GL_ARB_transform_feedback_overflow_query, GL_ARB_transpose_matrix,
    GL_ARB_uniform_buffer_object, GL_ARB_vertex_array_bgra,
    GL_ARB_vertex_array_object, GL_ARB_vertex_attrib_binding,
    GL_ARB_vertex_buffer_object, GL_ARB_vertex_program, GL_ARB_vertex_shader,
    GL_ARB_vertex_type_10f_11f_11f_rev, GL_ARB_vertex_type_2_10_10_10_rev,
    GL_ARB_viewport_array, GL_ARB_window_pos, GL_ATI_blend_equation_separate,
    GL_ATI_draw_buffers, GL_ATI_fragment_shader, GL_ATI_separate_stencil,
    GL_ATI_texture_compression_3dc, GL_ATI_texture_env_combine3,
    GL_ATI_texture_mirror_once, GL_EXT_abgr, GL_EXT_bgra, GL_EXT_blend_color,
    GL_EXT_blend_equation_separate, GL_EXT_blend_func_separate,
    GL_EXT_blend_minmax, GL_EXT_blend_subtract, GL_EXT_compiled_vertex_array,
    GL_EXT_copy_texture, GL_EXT_draw_buffers2, GL_EXT_draw_instanced,
    GL_EXT_draw_range_elements, GL_EXT_fog_coord, GL_EXT_framebuffer_blit,
    GL_EXT_framebuffer_multisample, GL_EXT_framebuffer_multisample_blit_scaled,
    GL_EXT_framebuffer_object, GL_EXT_framebuffer_sRGB,
    GL_EXT_gpu_program_parameters, GL_EXT_multi_draw_arrays,
    GL_EXT_packed_depth_stencil, GL_EXT_packed_float, GL_EXT_packed_pixels,
    GL_EXT_pixel_buffer_object, GL_EXT_point_parameters,
    GL_EXT_polygon_offset_clamp, GL_EXT_provoking_vertex,
    GL_EXT_rescale_normal, GL_EXT_secondary_color,
    GL_EXT_separate_specular_color, GL_EXT_shader_integer_mix,
    GL_EXT_shadow_funcs, GL_EXT_stencil_two_side, GL_EXT_stencil_wrap,
    GL_EXT_subtexture, GL_EXT_texture, GL_EXT_texture3D,
    GL_EXT_texture_array, GL_EXT_texture_compression_dxt1,
    GL_EXT_texture_compression_latc, GL_EXT_texture_compression_rgtc,
    GL_EXT_texture_compression_s3tc, GL_EXT_texture_cube_map,
    GL_EXT_texture_edge_clamp, GL_EXT_texture_env_add,
    GL_EXT_texture_env_combine, GL_EXT_texture_env_dot3,
    GL_EXT_texture_integer, GL_EXT_texture_lod_bias,
    GL_EXT_texture_mirror_clamp, GL_EXT_texture_object,
    GL_EXT_texture_rectangle, GL_EXT_texture_sRGB, GL_EXT_texture_sRGB_decode,
    GL_EXT_texture_shared_exponent, GL_EXT_texture_snorm,
    GL_EXT_texture_swizzle, GL_EXT_timer_query, GL_EXT_transform_feedback,
    GL_EXT_vertex_array, GL_EXT_vertex_array_bgra,
    GL_IBM_multimode_draw_arrays, GL_IBM_rasterpos_clip,
    GL_IBM_texture_mirrored_repeat, GL_INGR_blend_func_separate,
    GL_KHR_context_flush_control, GL_KHR_debug, GL_KHR_no_error,
    GL_KHR_texture_compression_astc_ldr,
    GL_KHR_texture_compression_astc_sliced_3d, GL_MESA_pack_invert,
    GL_MESA_shader_integer_functions, GL_MESA_texture_signed_rgba,
    GL_MESA_window_pos, GL_MESA_ycbcr_texture, GL_NV_blend_square,
    GL_NV_conditional_render, GL_NV_depth_clamp, GL_NV_fog_distance,
    GL_NV_light_max_exponent, GL_NV_packed_depth_stencil,
    GL_NV_primitive_restart, GL_NV_texgen_reflection,
    GL_NV_texture_env_combine4, GL_NV_texture_rectangle, GL_OES_EGL_image,
    GL_OES_read_format, GL_S3_s3tc, GL_SGIS_generate_mipmap,
    GL_SGIS_texture_border_clamp, GL_SGIS_texture_edge_clamp,
    GL_SGIS_texture_lod, GL_SUN_multi_draw_arrays

1 GLX Visuals
    visual  x   bf lv rg d st  colorbuffer  sr ax dp st accumbuffer  ms  cav
  id dep cl sp  sz l  ci b ro  r  g  b  a F gb bf th cl  r  g  b  a ns b eat
----------------------------------------------------------------------------
0x075 24 tc  0  24  0 r  y .   8  8  8  0 .  .  0 16  0  0  0  0  0  0 0 None

1 GLXFBConfigs:
    visual  x   bf lv rg d st  colorbuffer  sr ax dp st accumbuffer  ms  cav
  id dep cl sp  sz l  ci b ro  r  g  b  a F gb bf th cl  r  g  b  a ns b eat
----------------------------------------------------------------------------
0x075 24 tc  0  24  0 r  y .   8  8  8  0 .  .  0 16  0  0  0  0  0  0 0 None
```

**Comment by dmd** [comment]
setting `LIBGL_ALWAYS_SOFTWARE=1` doesn't help.

setting `LIBGL_ALWAYS_INDIRECT=1` doesn't work, because x2go uses OpenGL 1.2 not 1.4 (which could be a clue here?)

setting `GDK_BACKEND=x11` doesn't help.

**Comment by pauldmccarthy** [comment]
So the real error is this, which suggests that something has gone wrong with GL initialisation:

```wx._core.wxAssertionError: C++ assertion ""Assert failure"" failed at /home/conda/feedstock_root/build_artifacts/wxpython_1616628589829/work/ext/wxWidgets/src/unix/glx11.cpp(626) in InitVisual(): Failed to get a XVisualInfo for the requested attributes.```

I have limited experience with nomachine/x2go, but do have a hack in place when initialising GL:

https://git.fmrib.ox.ac.uk/fsl/fsleyes/fsleyes/-/blob/master/fsleyes/gl/__init__.py#L794-818

So I'm wondering if your issue can be avoided by doing something similar (and hoping that the default depth/stencil buffers are sufficient). Would you be able to modify `/cm/shared/fsl-6.0.4/fslpython/envs/fslpython/lib/python3.7/site-packages/fsleyes/gl/wxglslicecanvas.py`, changing the line:

```
wxgl.GLCanvas          .__init__(self, parent, **attrs)
```

to:

```
wxgl.GLCanvas          .__init__(self, parent)
```

And see if that resolves the problem?

**Comment by dmd** [comment]
I changed:

```python
wxgl.GLCanvas          .__init__(self, parent, **attrs)
```

to:

```python
try:
    wxgl.GLCanvas          .__init__(self, parent, **attrs)
except Exception:
    wxgl.GLCanvas          .__init__(self, parent)
```

and that appears to fix the problem!

Awesome. Can you update this ticket when this change is merged?

**Comment by pauldmccarthy** [comment]
Great! I'll try and have this fix out by next week (am a little busy right now) ...

**Comment by pauldmccarthy** [comment]
@dmd sorry, I forgot to ping you - the fix for this issue was added in version 1.0.11 (14th June). Feel free to re-open this issue if you're still having problems
